### ISSUE TITLE
Lack of provider method/field in class besom.api.eks.Cluster

### ISSUE THREAD

**Issue by polkx** [VirtusLab/besom#397]
When I ask pulumi AI "aws eks hello world” I get:
https://www.pulumi.com/ai/conversations/6ff221c6-9b0b-49be-8233-1afaf3e7fea7
```ts
import * as aws from "@pulumi/aws";
import * as eks from "@pulumi/eks";

// Create an EKS cluster with the default configuration.
const cluster = new eks.Cluster("my-cluster");

// Define the "hello world" Deployment.
const appLabels = { app: "hello-world" };
const deployment = new k8s.apps.v1.Deployment("hello-world-deployment", {
    metadata: { labels: appLabels },
    spec: {
        replicas: 2,
        selector: { matchLabels: appLabels },
        template: {
            metadata: { labels: appLabels },
            spec: {
                containers: [{
                    name: "hello-world",
                    image: "paulbouwer/hello-kubernetes:1.5",
                    ports: [{ containerPort: 8080 }],
                }],
            },
        },
    },
}, { provider: cluster.provider });

// Create a LoadBalancer Service to expose the "hello world" Deployment.
const service = new k8s.core.v1.Service("hello-world-service", {
    metadata: { labels: appLabels },
    spec: {
        type: "LoadBalancer",
        selector: appLabels,
        ports: [{ port: 80, targetPort: 8080 }],
    },
}, { provider: cluster.provider });

// Export the URL for the load-balanced service.
export const url = service.status.loadBalancer.ingress[0].hostname;
```

I want to use 
```
{ provider: cluster.provider } 
```
in Besom. Like below:
```
opts = opts(provider = cluster.provider)
```
But class besom.api.eks.Cluster don't have a provider method.

**Comment by lbialy** [comment]
First of all, is cluster.provider a field on the resource or a core method exposing provider used to deploy this particular resource? I assume the latter and that in other implementations it's just a method inherited from Resource, right? If so, it should be trivial to add it in core. All it has to do is to reach to the guts of context and fetch the provider from the state. The issue is - there's a probability that field `provider` exists somewhere so I guess we should do this as an extension method, which won't cause conflicts.

**Comment by lbialy** [comment]
@pawelprazak your opinion here?

**Comment by lbialy** [comment]
You're thinking in terms of inheritance. We are free to ditch inheritance whenever we want to! This. is. Scala!

We almost never upcast to `Resource` supertype so the easiest way to get this is to just use extension methods and the design is kinda trivial:
```
extension (cr: CustomResource) // this also handles ProviderResources but for them it's probably nonsensical
  def provider: Output[ProviderResource] = ???
  
extension (cb: ComponentResource)
  def providers: Output[Map[String, ProviderResource]] = ???
  
extension (cb: RemoteComponentResource)
  def providers: Output[Map[String, ProviderResource]] = ???
```

**Comment by lbialy** [comment]
fixed with #397 @polkx can you verify it works like you'd expect via `just clean-all publish-local-all` with 0.4.0-SNAPSHOT version of core?
